### PR TITLE
Allow to set organization visibility (public, internal, private) 

### PR DESCRIPTION
**Pull request by DblK**
This will implement a way to have a visibility on organization (Public, Limited & Private).
It will implement #714 .

The UI modification is in settings:
image

The rules for displaying organization and repositories within organization are described in #714

### CODE REVIEW

**Comment by DblK** [comment]
UI modifications in Admin:
image

Finally both Explore/Organizations and Explore/Repositories are filtered to display the right organizations.

**Comment by camlafit** [comment]
Hi

Looks great but :)
* we have a problem with administrator account, it can manage organization policy from **admin/orgs** but lost public access, it can see all private organization from *explore/organizations** . It looks logical but seems strange behavior. I don't have any idea to improve this behavior.
*  from **explore/organizations** page , I get this error going to any organization page : 
```
template: explore/repo_list:3:21: executing "explore/repo_list" at <.Owner.Visibility>: can't evaluate field Visibility in type *models.User
```
* on **admin/orgs** miss an icon status as done in **explore/organizations**
* all user are displayed on **explore/users** without user connected
* Default organization status is OK, a new organization take service.DEFAULT_VISIBILITY value

**Comment by camlafit** [comment]
Hi 

Looks also a problem about displaying repo and orgnizaiton. An user set as repo owner can't see neither orgnaization page, neither repo page (404).
On another organization set also as private it's ok. On dashboard user all commits are displayed.

log show this : 
```
2017/05/22 17:28:49 [...ules/context/repo.go:527 func1()] [E] CheckUnit: You have not allowed to visit this repository unit: 1
2017/05/22 17:28:52 [...routers/repo/view.go:270 Home()] [E] Home: Cannot find any unit allowed on this repository
2017/05/22 17:28:54 [...ules/context/repo.go:527 func1()] [E] CheckUnit: You have not allowed to visit this repository unit: 1
```

**Comment by DblK** [comment]
Thanks @camlafit for your tests.
So here are my answers:

1. That's true that admin can see everything on *explore/organizations*. If I understand what you said, you wish admin can be both admin and public? In others words, you wish admin can check if he wants standard user view or not? If yes, I think admin should stay admin. I don't see how to change that. Maybe a new user settings for admin only to display as "non admin user"? If yes, another PR will be done for that because I think it's an enhancement of what is done here.

2. I saw the issue. It is not happening when going an organization has no repository. I saw the bug and I will fix it.
**EDIT!** Fixed in next commit

3. I did not change this page so it is normal. I will add the icon for *private* organizations.
**EDIT:** Here are the change done to *admin/orgs* and *admin/repos*
image
image


4. I think this should be made in another PR. Displaying all users are not part of this feature. Please made another issue and I will try to fix it too. Please explain what you want to have. I bet you wish to have something like a *private* or *limited* user.

5. Glad to hear it is working. It was not on your issue but I thought it can be useful.

6. So, the user is not the creator of the organization but has been part of the Owner team after that? Could you explain more about the two different use case so I can reproduce and fix it before merging.

7. I will review the dashboard to ensure nothing appears. Or I think it is working as expected but when user left the organization they still see in the dashboard. If yes, it is not related to this PR but should be done as another issue and I will do the PR.

If you have other remarks, feel free to add more. I will dig into the first ones you gave me ;)

**Comment by camlafit** [comment]
Hi

Thanks for improvement

1/ As admin we can see all organization from **explore/organizations** andb loooks ok. But all link target a 404 this part doesn't look normal. We should find a 403 forbidden page or a redirection to admin organization page. 
Now about organization page look ok, empty or not page is displayed. But link to repo target a 404.

2/ looks solved :)

3/ great looks better as this 👍 

4/ yes looks better to split #714 I've opened #1780 about user displaying

5/ yes it's useful. On my case all organization are private with unique exception. Easier as this.

6/ dashboard user looks updated, organization with pb is no more displayed.

7/ I've an organization set as Private, with an unique team "owner" and unique member and unique repo. On his dashboard commits are displayed, but it's not possible to go to their repository, 404 page.
Configuration looks same as others organizations but I don't understand why I get this 404.
Maybe in this case user was creator and not set after as owner. I don't remember.

I've two accounts , an administrator and a personal. In my process I prepare organization and repo from admin account. I add standard account as owner
Looks possible I've create only organization in this case and create repo from normal user. I don't remember. It's possible also to create repo in personnal space and move after to organization. Sorry I don't remember. I don't know if I see history change about this orga/repo.

**Comment by DblK** [comment]
I could not see the problem with 1) nor 7).
From my point of view, unless you can provide me a use case to reproduce it each time, this PR is done and waiting for the review of code.

**Comment by DblK** [comment]
According to our discussion (@camlafit) and my investigation, the error is not part of this PR (see #1794).
This PR is good to go ^^

**Comment by DblK** [comment]
For the change in the UI in *explore/repo* to add the lock close to private organization:
image

**Comment by camlafit** [comment]
Hi

Looks again a pb. If user is only organization member, repositories are not lister.
Repositories are displayed only if user is set in owner team.

**Comment by DblK** [comment]
Your problem is not specific to this PR but to the master. Look like you should add a comment for #1794.
This is related to the implementation of Units. No repositories are show but public.

Please test it also on master to be sure the bug is from this PR. If it is not working on master, please fill a new issue or complete an existing one.

So for me, so far no bug from this PR (some but inherited from master).

**Comment by lunny** [comment]
need rebase and more tests

**Comment by manuelmanhart** [comment]
Hi.

Can we do anything to help get this merged back into master? At least would like / need to have that feature very much. But I don't see whats the missing / open issues here.

Regards, Manuel

**Comment by DblK** [comment]
If somebody can do the rebase and add tests, go for it. I which it can be merged as well.

**Comment by pragmaware** [comment]
+1 for this feature

**Comment by techknowlogick** [comment]
I can do the rebase, if someone wants to do the tests.

@DblK Even though I'm a maintainer, I don't have "owner" permissions and can't edit your PR. Could you grant me those directly?

**Comment by camlafit** [comment]
Hello

As I've asked to this feature :) I can do some tests.
Only need to give me commit to fetch and I can build in my forge.

**Comment by lunny** [comment]
@techknowlogick need @DblK's permission.

File: modules/setting/setting.go
Comment (bkcsoft) [inline modules/setting/setting.go]: This list shouldn't be hard-coded but fetched from the hashmap above. If there's no go-func in stdlib to do that, use this:

```go
func ExtractKeysFromMapString(in map[string]interface{}) (keys []string) {
  for k, _ := range in {
    keys = append(keys, k)
  }
}
```

File: models/user.go
Comment (bkcsoft) [inline models/user.go]: call this `IsOrganizationMemeber` instead. Since `User` is redundant here.

File: options/locale/locale_en-US.ini
Comment (bkcsoft) [inline options/locale/locale_en-US.ini]: what is a "connected" user?

File: models/org.go
Comment (bkcsoft) [inline models/org.go]: One is not able to set it when it's being created? What if the default is "public" and you want it "private"? 🤔

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by camlafit** [comment]
Hello

I've tried to rebase pr on master.
I get a conflct on options/locale/locale_en-US.ini
```
[units]
<<<<<<< HEAD
error.no_unit_allowed_repo = Cannot find any unit on this repository which
you are allowed to access
error.unit_not_allowed = You are not allowed to visit this repository unit
=======
error.no_unit_allowed_repo = Cannot find any unit allowed on this repository
error.unit_not_allowed = You have not allowed to visit this repository unit
>>>>>>> Add UI visibility in organization settings

```

I get some others conflicts on patch "Implement global configuration for
new organizations" on these files
```
  custom/conf/app.ini.sample
  models/org.go
  modules/setting/setting.go
  options/locale/locale_en-US.ini
  templates/admin/config.tmpl

```
I can't resolve them.

File: models/org.go
Comment (DblK) [reply models/org.go]: You can still change it after anyway. I do not remember this code. Need deeper look

File: options/locale/locale_en-US.ini
Comment (bkcsoft) [reply options/locale/locale_en-US.ini]: Then it should say "Visible to logged in users only" to be consistent :)

**Comment by DblK** [comment]
@techknowlogick you should have the right, it's checked. Tell me if you still have trouble.

**Comment by scr34m** [comment]
A very really missing feature, any way i can help / contribute to be merged?

**Comment by techknowlogick** [comment]
Ok. I've just pushed a commit that resolves all the conflicts.

File: models/org.go
Comment (lunny) [inline models/org.go]: 2018

File: templates/explore/repo_list.tmpl
Comment (scr34m) [inline templates/explore/repo_list.tmpl]: An extra test required because .Owner is not defined then an error occur ` <.Owner.Visibility>: can't evaluate field Visibility in type *models.User`

Possible solution:

```
{{else if .Owner}}
    {{if eq .Owner.Visibility 3}}
        <span class="text gold"><i class="octicon octicon-lock"></i></span>
    {{end}}
{{end}}
````

Comment (techknowlogick) [reply templates/explore/repo_list.tmpl]: Thanks @scr34m I've just updated the template

**Comment by tarelda** [comment]
If merged when this feature will be available in release? Really looking forward to migrate my gitlab (I found out that it is eating around 16GB of RAM) installation to gitea, but I need this feature.

**Comment by herrbischoff** [comment]
Any updates to share on this? I'm also looking forward to switch to Gitea from GitLab for project work. Clients should only ever see the organizations and their users of which they're part of.

**Comment by techknowlogick** [comment]
~Manual testing note: I have been able to access the profile pages of private orgs if I have the URL directly. This is something that I'll need to resolve.~

Fixed

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: @bkcsoft fixed. User can now chose visibility when creating org.

**Comment by herrbischoff** [comment]
Is there anything apart from the missing tests that keeps this from getting merged?

**Comment by techknowlogick** [comment]
@herrbischoff per status above, tests are the only item remaining.

**Comment by techknowlogick** [comment]
ok. Tests are done.

File: templates/explore/organizations.tmpl
Comment (lafriks) [inline templates/explore/organizations.tmpl]: This will break paging as if most organizations are private can be that you will get empty pages or pages with very few orgs in it. These conditions should be already part of database query

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lafriks** [comment]
Also I might have missed but I did not see were API was limited to only return allowed orgs

File: templates/explore/organizations.tmpl
Comment (DblK) [reply templates/explore/organizations.tmpl]: Good catch, when I did that I did not have that many user so did not see the pages. Everything was one page. Agree with your solution.

Comment (techknowlogick) [reply templates/explore/organizations.tmpl]: @DblK I've just resolved the latest conflicts. I'm going on vacation until mid-august, do you want to take over this PR again, and add in the SQL conditions?

**Review by bkcsoft — DISMISSED** [review]

File: models/user.go
Comment (DblK) [reply models/user.go]: `IsUserPartOfOrg` should have been better

**Comment by lunny** [comment]
This could be easy to implement after #5314 merged.

**Comment by lunny** [comment]
I'm afraid `VisibleTypeLimited` is not really used in this PR, maybe remove it currently to make code easier.

**Comment by techknowlogick** [comment]
@lunny I think it is still needed, because even though it isn't isn't used directly, there are conditionals that match it.

**Comment by pragmaware** [comment]
Any followup on this?
Private organisations are a very useful (and I see, widely requested) feature.

**Comment by techknowlogick** [comment]
@bkcsoft @lunny @lafriks ready for review.

File: modules/setting/setting.go
Comment (techknowlogick) [reply modules/setting/setting.go]: Done

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: done

File: models/migrations/migrations.go
Comment (adelowo) [inline models/migrations/migrations.go:2]: @techknowlogick 2019

File: routers/org/org.go
Comment (adelowo) [inline routers/org/org.go:2]: @techknowlogick 2019

File: models/migrations/migrations.go
Comment (techknowlogick) [reply models/migrations/migrations.go:2]: This one was kept as 2017 because it was changed then, but the copyright was missed in other PRs.

File: routers/org/org.go
Comment (techknowlogick) [reply routers/org/org.go:2]: This file was modified in 2018, but the copyright wasn't added.

File: models/org.go
Comment (kolaente) [inline models/org.go:399]: Won't this return true no matter if the user is part of the org or not? (Because the default case is `true`)

File: models/org.go
Comment (adelowo) [inline models/org.go]:
```suggestion
// HasOrgVisible tells if the given user can see one or more of the given orgs
```

File: models/user.go
Comment (adelowo) [inline models/user.go]:
```suggestion
// VisibleType defines the visibility (Organization only)
```

**Review by kolaente — COMMENTED** [review]
Also what about adding this to the comparison?

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: Thanks :) Done.

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: Thanks :) Done.

File: models/org.go
Comment (techknowlogick) [reply models/org.go:399]: Not necessarily. Because if org is private then first case will match, but in the first case if user is not part of org switch case will complete and below (Line 400) will return false.

Comment (kolaente) [reply models/org.go:399]: Then why do we need this switch case if we could also do this (could we?) with something like 
```go
if org.Visibility == VisibleTypePrivate && org.IsUserPartOfOrg(user.ID) {
	return true
}

Comment (techknowlogick) [reply models/org.go:399]: As we need to handle the other types of visibility, so the complete conditional would be:

```go
if (org.Visibility == VisibleTypePrivate && org.IsUserPartOfOrg(user.ID)) || org.Visibility == VisibleTypeLimited || org.Visibility == VisibleTypePublic {
	return true
}
```
the above switch case is cleaner

Comment (kolaente) [reply models/org.go:399]: But where are we explicitly checking these cases? They're not in the switch case, so if `org.Visibility` is either `VisibleTypeLimited` or `VisibleTypePublic` it will return true. IMHO if we're not exiplicitly checking them, it wouldn't make a difference if we included them into a shorter `if`...

Sorry if I'm picking about obvious things here, but I think I don't quite understand your thought.

Comment (techknowlogick) [reply models/org.go:399]: We are checking for both of those visibility types in the switch, they are caught by the `default` case.

File: models/user.go
Comment (lafriks) [inline models/user.go]: Please reuse constants `VisibleTypePublic` etc here

File: models/user.go
Comment (lafriks) [inline models/user.go]: Same here, please, reuse constants `VisibleTypePublic`...

File: models/user.go
Comment (lafriks) [inline models/user.go]: same here

File: modules/setting/setting.go
Comment (lafriks) [inline modules/setting/setting.go]: reuse constants also here.. Also are they are correct? (in code in model it is used as 0, 1, 2 - might be it is incorrect in selects...)

File: templates/explore/organizations.tmpl
Comment (lafriks) [inline templates/explore/organizations.tmpl]: I don't think it us good idea to filter out user list in view... it should be done in select to get only needed data, otherwise paging will be wrong and could be that you get 10 pages with one or two orgs per page

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/setting/setting.go
Comment (techknowlogick) [reply modules/setting/setting.go]: Can't import constants into settings because cyclical (file with constants already imports settings). They are correct, the reason why 0 is used in select is because users (who don't have visibility) are set to 0 and org/user search uses the same code because they are essentially the same. So I have to make sure that if a code needs to search for users then users are included.

File: templates/explore/organizations.tmpl
Comment (techknowlogick) [reply templates/explore/organizations.tmpl]: paging won't be wrong because SQL matches this case, so the results passed into template is already filtered. I will remove this conditional though, because it is not needed.

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: Done

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: Done

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: Done

File: templates/explore/organizations.tmpl
Comment (techknowlogick) [reply templates/explore/organizations.tmpl]: Updated

File: models/user.go
Comment (lunny) [inline models/user.go:1356]: VisibleType?

File: models/user.go
Comment (lunny) [inline models/user.go:1395]: ```var accessCond = builder.Or(...```

File: modules/setting/setting.go
Comment (lunny) [reply modules/setting/setting.go]: Maybe we should move all structs to `models/structs` as a sub package.

File: models/org.go
Comment (lunny) [inline models/org.go]: We don't need `orgs` but only `org` or we should return`[]bool`.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: Done. This was used to see if any of the orgs that a user belongs to can be seen when viewing the profile page. There is additional logic in the template to hide org if user shouldn't see it.

File: models/user.go
Comment (techknowlogick) [reply models/user.go:1395]: done

File: models/user.go
Comment (techknowlogick) [reply models/user.go:1356]: I used `Private` to match the options in `SearchRepoOptions`

File: modules/setting/setting.go
Comment (techknowlogick) [reply modules/setting/setting.go]: done.

File: models/user.go
Comment (lafriks) [reply models/user.go]: But migration will set visibility to 1 for users also...

File: models/user.go
Comment (techknowlogick) [reply models/user.go:1395]: This fails on MSSQL, so I will revert

**Comment by techknowlogick** [comment]
From some more testing I found some areas where I still need to improve with SQL statements. I've marked this as WIP again.

**Comment by techknowlogick** [comment]
ok. no longer WIP anymore. ready to get reviewed again.

File: models/org.go
Comment (zeripath) [inline models/org.go]: I think this `switch` block is confusing. Would a two case `if` be clearer?

File: models/user.go
Comment (zeripath) [inline models/user.go]: Are we starting at 1 here to allow old organisations to be undefined?

File: models/user.go
Comment (zeripath) [inline models/user.go:532]: Hmm. I'm confused by this function signature & comment.

I guess that u is meant to be the organisation, and you're checking that user ID is in this organisation. 

I'm not sure what the name of this function should be to completely ensure it's clear but you should definitely change the comment saying: returns true is user ID is part of the u organisation.

I also have to admit that I have a little pet hate for functions that hide database access like this - it's too easy to miss in a transaction.

File: models/user.go
Comment (zeripath) [reply models/user.go:1356]: Hmm I think we're going to immediately get an issue opened with a user complaining that they want to search on the different types of visibility - is it really disruptive to change it or do you think we should deal with that when the inevitable occurs?

**Review by zeripath — COMMENTED** [review]
Just some comments - not a formal nack.

Thanks @techknowlogick for doing so much on this.

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: The alternative is:
```go
if (org.Visibility == VisibleTypePrivate && org.IsUserPartOfOrg(user.ID)) || org.Visibility == VisibleTypeLimited || org.Visibility == VisibleTypePublic {
	return true
}
return false
```
as such I prefer the switch case.

File: models/user.go
Comment (techknowlogick) [reply models/user.go]: Updated to start at 0.

File: models/user.go
Comment (techknowlogick) [reply models/user.go:532]: I tried to be clever and change u to org, however lint (rightfully so), error'd out. So I've updated comment to match your suggestion.

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: To add to this, there are conflicting reviews on this switch case, and I am unable to satisfy all of them.

File: models/user.go
Comment (techknowlogick) [reply models/user.go:1356]: I think we can add this in later with a slice for VisibleTypes and have a conditional (after the other conditionals) that says `visibility in ([]VisibilityTypes)`, that way it'll filter on the already filtered list to ensure that a org that the user can't see isn't accidentally shown to user.

I am hesitant to add this in now, and want to get the minimal feature set out for this before adding anything else.

File: models/org.go
Comment (zeripath) [reply models/org.go]: I can imagine - the bigger the patch the more the bikeshedding... sorry

I was thinking instead of:
```go
switch org.Visibility {
	case VisibleTypePrivate:
		if org.IsUserPartOfOrg(user.ID) {
			return true
		}
	default:
		return true
	}
	return false
}
```

do this:
```go
if org.Visibility == VisibleTypePrivate {
	if org.IsUserPartOfOrg(user.ID) {
		return true
	}
	return false
}
return true
```

File: models/user.go
Comment (zeripath) [reply models/user.go:532]: Thanks!

Comment (zeripath) [reply models/user.go:532]: Oh change it to the `user with id userID is part of the organization, u.` or `userID is part of the organization`. It's still slightly unclear if you're just grepping the source code.

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: Ah, perfect. I think that satisfies all the reviews. I will update to your suggestion. Thanks :)

File: models/user.go
Comment (techknowlogick) [reply models/user.go:532]: oops. Thanks for checking again after the first update. Hopefully I got it right this time.

File: models/org.go
Comment (techknowlogick) [reply models/org.go]: Done.

File: models/user.go
Comment (zeripath) [reply models/user.go:532]: Cools! Thanks dude! I've just re-hit the extremely confusing similarly named functions:

https://github.com/go-gitea/gitea/blob/master/models/ssh_key.go#L765

&

https://github.com/go-gitea/gitea/blob/master/models/ssh_key.go#L777

**Comment by camlafit** [comment]
Hello

Thanks a lot to your great job about this query. No problem to test this feature on mine production instance.
When you think could be testable and updatable, don't forget to ping me :) I use mysql instance.

File: routers/api/v1/org/org.go
Comment (lafriks) [inline routers/api/v1/org/org.go]: no need for variable here

File: routers/api/v1/repo/repo.go
Comment (lafriks) [inline routers/api/v1/repo/repo.go]: no need for variable here

File: routers/repo/view.go
Comment (lafriks) [inline routers/repo/view.go]: no need for variable here

File: routers/user/home.go
Comment (lafriks) [inline routers/user/home.go]: no need for variable here

**Review by lafriks — COMMENTED** [review]
just nit :)

File: routers/user/home.go
Comment (techknowlogick) [reply routers/user/home.go]: done

File: routers/repo/view.go
Comment (techknowlogick) [reply routers/repo/view.go]: done

File: routers/api/v1/repo/repo.go
Comment (techknowlogick) [reply routers/api/v1/repo/repo.go]: done

File: routers/api/v1/org/org.go
Comment (techknowlogick) [reply routers/api/v1/org/org.go]: done

**Review by lafriks — APPROVED** [review]

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@techknowlogick a new idea, maybe `modules/structs/structs.go` could be renamed to `modules/structs/org_type.go` then different structs have different files.

**Comment by techknowlogick** [comment]
@lunny good idea, I've just renamed `modules/structs/structs.go` to your suggestion.

File: models/repo_list.go
Comment (lunny) [inline models/repo_list.go]: Please break the line before `)`, you can place a `,` at the end of the line.

Comment (techknowlogick) [reply models/repo_list.go]: Thanks 😄 

updated.

**Comment by techknowlogick** [comment]
@lunny ok, successful build after all changes you requested.

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/setting/setting.go
Comment (techknowlogick) [reply modules/setting/setting.go]: updated

File: models/org.go
Comment (lunny) [reply models/org.go:387]: Maybe you missed something?

Comment (techknowlogick) [reply models/org.go:387]: If we returned false that means that the user shouldn't see public orgs, so I simplified the if statement.

Comment (techknowlogick) [reply models/org.go:387]: the only time it should ever be false is if private, and the user doesn't belong to org

**Comment by techknowlogick** [comment]
@lunny resolved conflict with recent settings module change. please review.

**Review by lunny — APPROVED** [review]

**Comment by DblK** [comment]
Thx @techknowlogick for finishing/rebasing my PR. I will love to use this feature.
Finally I will be able to open my own Gitea instance to friends :p

**Comment by nicorac** [comment]
Just upgraded to [`64ce159`](/go-gitea/gitea/commit/64ce159a6eacc81962d07a8f5ef7f69c17365363) and noticed an issue in Organizations list admin page (/admin/orgs).

Opened an issue here: #6111

**Comment by techknowlogick** [comment]
@DblK 😄 Thanks for starting the PR.

**Comment by DblK** [comment]
Good work @techknowlogick. Finally landed in a release ;)

**Comment by camlafit** [comment]
Works as a charm \o/

**Pull request merged**
